Augmentation in `nobrainer.dataset.get_dataset` results in internal tensorflow exception

Command-line output of `nobrainer info`.
```
(env) Singularity> nobrainer info
Python:
 Version: 3.8.10
 Implementation: CPython
 64-bit: True
 Packages:
  Nobrainer: 0+unknown
  Nibabel: 5.1.0
  Numpy: 1.23.5
  TensorFlow: 2.12.0
   GPU support: True
   GPU available: True
  TensorFlow-Probability: 0.20.1

System:
 OSType: Linux
 Release: 4.18.0-372.9.1.el8.x86_64
 Version: #1 SMP Tue May 10 14:48:47 UTC 2022
 Architecture: x86_64

Timestamp: 2023/07/17 23:16:30
```

What were you trying to do?
Augment volumetric data by adding gaussian noise

What did you expect will happen?
The data will be augmented with gaussian noise

What actually happened?
The call to `get_dataset` failed with a traceback internal to `tensorflow.cond()`

```
import nobrainer

filepaths = nobrainer.io.read_csv(nobrainer.utils.get_data())
nobrainer.tfrecord.write(
    features_labels=filepaths,
    filename_template='data/tmp/data*.tfrec',
    examples_per_shard=10,
)

from nobrainer.intensity_transforms import addGaussianNoise
nobrainer.dataset.get_dataset(
    file_pattern='data/tmp/data*.tfrec',
    augment=[(addGaussianNoise, {'noise_mean': 0.1, 'noise_std': 0.5})],
    n_classes=1,
    batch_size=1,
    volume_shape=(256, 256, 256),
)
```

Traceback:
```
Traceback (most recent call last):
  File "issues/aug.py", line 13, in <module>
    nobrainer.dataset.get_dataset(
  File "/om/weka/gablab/ohinds/projects/nobrainer/nobrainer/dataset.py", line 154, in get_datas
et
    dataset = dataset.map(
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/py
thon/data/ops/dataset_ops.py", line 2240, in map
    return map_op._map_v2(
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/py
thon/data/ops/map_op.py", line 40, in _map_v2
    return _ParallelMapDataset(
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/py
thon/data/ops/map_op.py", line 148, in __init__
    self._map_func = structured_function.StructuredFunctionWrapper(
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/py
thon/data/ops/structured_function.py", line 261, in __init__
    self._function = fn_factory()
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/py
thon/eager/polymorphic_function/tracing_compiler.py", line 232, in get_concrete_function
    concrete_function = self._get_concrete_function_garbage_collected(
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/py
thon/eager/polymorphic_function/tracing_compiler.py", line 202, in _get_concrete_function_garba
ge_collected
    concrete_function, _ = self._maybe_define_concrete_function(args, kwargs)
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/py
thon/eager/polymorphic_function/tracing_compiler.py", line 166, in _maybe_define_concrete_funct
ion
    return self._maybe_define_function(args, kwargs)
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/py
thon/eager/polymorphic_function/tracing_compiler.py", line 396, in _maybe_define_function
    concrete_function = self._create_concrete_function(
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/python/eager/polymorphic_function/tracing_compiler.py", line 300, in _create_concrete_function
    func_graph_module.func_graph_from_py_func(
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/python/framework/func_graph.py", line 1214, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/python/data/ops/structured_function.py", line 238, in wrapped_fn
    ret = wrapper_helper(*args)
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/python/data/ops/structured_function.py", line 169, in wrapper_helper
    ret = autograph.tf_convert(self._func, ag_ctx)(*nested_args)
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py", line 692, in wrapper
    raise e.ag_error_metadata.to_exception(e)
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py", line 689, in wrapper
    return converted_call(f, args, kwargs, options=options)
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py", line 439, in converted_call
    result = converted_f(*effective_args, **kwargs)
  File "/tmp/__autograph_generated_filex8uwlhry.py", line 7, in <lambda>
    tf__lam = (lambda x, y: ag__.with_function_scope((lambda lscope: ag__.converted_call(tf.cond, ((ag__.converted_call(tf.random.uniform, ((1,),), None, lscope) > 0.5),), dict(true_fn=ag__.autograph_artifact((lambda : ag__.converted_call(transform, (x, y), dict(**kwargs), lscope))), false_fn=ag__.autograph_artifact((lambda : (x, y)))), lscope)), 'lscope', ag__.STD))
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/python/autograph/core/function_wrappers.py", line 113, in with_function_scope
    return thunk(scope)
  File "/tmp/__autograph_generated_filex8uwlhry.py", line 7, in <lambda>
    tf__lam = (lambda x, y: ag__.with_function_scope((lambda lscope: ag__.converted_call(tf.cond, ((ag__.converted_call(tf.random.uniform, ((1,),), None, lscope) > 0.5),), dict(true_fn=ag__.autograph_artifact((lambda : ag__.converted_call(transform, (x, y), dict(**kwargs), lscope))), false_fn=ag__.autograph_artifact((lambda : (x, y)))), lscope)), 'lscope', ag__.STD))
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py", line 377, in converted_call
    return _call_unconverted(f, args, kwargs, options)
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py", line 458, in _call_unconverted
    return f(*args, **kwargs)
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py", line 153, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/om/weka/gablab/ohinds/projects/nobrainer/env/lib/python3.8/site-packages/tensorflow/python/ops/cond_v2.py", line 684, in _make_indexed_slices_indices_types_match
    assert len(set(outs_per_branch)) == 1, outs_per_branch
AssertionError: in user code:

    File "/om/weka/gablab/ohinds/projects/nobrainer/nobrainer/dataset.py", line 155, in None  *
        )

    AssertionError: [1, 2]

```

--- a/nobrainer/intensity_transforms.py
+++ b/nobrainer/intensity_transforms.py
@@ -43,8 +43,9 @@ def addGaussianNoise(x, y=None, trans_xy=False, noise_mean=0.0, noise_std=0.1):
 
         y = tf.cast(y, tf.float32)
         return tf.math.add(x, noise), tf.math.add(y, noise)
-    else:
+    if y is None:
         return tf.math.add(x, noise)
+    return tf.math.add(x, noise), y
 
 
 def minmaxIntensityScaling(x, y=None, trans_xy=False):
@@ -90,6 +91,8 @@ def minmaxIntensityScaling(x, y=None, trans_xy=False):
         ymin = tf.cast(tf.reduce_min(y), tf.float32)
         ymax = tf.cast(tf.reduce_max(y), tf.float32)
         y = tf.divide(tf.subtract(y, ymin), tf.add(tf.subtract(ymax, ymin), ep))
+    if y is None:
+        return x
     return x, y
 
 
@@ -151,9 +154,9 @@ def customIntensityScaling(x, y=None, trans_xy=False, scale_x=[0.0, 1.0], scale_
         )
         diff_y = tf.subtract(maxy, miny)
         y = tf.add(tf.multiply(y_norm, diff_y), miny)
-        return x, y
-    else:
+    if y is None:
         return x
+    return x, y
 
 
 def intensityMasking(x, mask_x, y=None, trans_xy=False, mask_y=None):
@@ -206,8 +209,9 @@ def intensityMasking(x, mask_x, y=None, trans_xy=False, mask_y=None):
         if mask_y.shape[0] != y.shape[0] and mask_x.shape[1] != x.shape[1]:
             raise ValueError("Label Masks shape should be same as Label")
         return x, tf.multiply(y, mask_y)
-    else:
+    if y is None:
         return x
+    return x, y
 
 
 def contrastAdjust(x, y=None, trans_xy=False, gamma=1.0):
@@ -265,5 +269,6 @@ def contrastAdjust(x, y=None, trans_xy=False, gamma=1.0):
         y = tf.pow(tf.divide(tf.subtract(y, ymin), tf.add(y_range, ep)), gamma)
         y = tf.add(tf.multiply(y, y_range), ymin)
         return x, y
-    else:
+    if y is None:
         return x
+    return x, y

--- a/nobrainer/spatial_transforms.py
+++ b/nobrainer/spatial_transforms.py
@@ -46,9 +46,9 @@ def centercrop(x, y=None, finesize=64, trans_xy=False):
             y = tf.convert_to_tensor(y)
         y = tf.cast(y, tf.float32)
         y = y[y1 : y1 + th, x1 : x1 + tw, :]
-        return x, y
-    else:
+    if y is None:
         return x
+    return x, y
 
 
 def spatialConstantPadding(x, y=None, trans_xy=False, padding_zyx=[1, 1, 1]):
@@ -98,9 +98,9 @@ def spatialConstantPadding(x, y=None, trans_xy=False, padding_zyx=[1, 1, 1]):
             y = tf.convert_to_tensor(y)
         y = tf.cast(y, tf.float32)
         y = tf.pad(y, padding, "CONSTANT")
-        return x, y
-    else:
+    if y is None:
         return x
+    return x, y
 
 
 def randomCrop(x, y=None, trans_xy=False, cropsize=16):
@@ -142,8 +142,9 @@ def randomCrop(x, y=None, trans_xy=False, cropsize=16):
         stacked = tf.stack([x, y], axis=0)
         cropped = tf.image.random_crop(stacked, [2, cropsize, cropsize, x.shape[2]])
         return cropped[0], cropped[1]
-    else:
+    if y is None:
         return tf.image.random_crop(x, [cropsize, cropsize, x.shape[2]])
+    return tf.image.random_crop(x, [cropsize, cropsize, x.shape[2]]), y
 
 
 def resize(x, y=None, trans_xy=False, size=[32, 32], mode="bicubic"):
@@ -189,9 +190,9 @@ def resize(x, y=None, trans_xy=False, size=[32, 32], mode="bicubic"):
             y = tf.convert_to_tensor(y)
         y = tf.cast(y, tf.float32)
         y = tf.image.resize(y, size, method=mode)
-        return x, y
-    else:
+    if y is None:
         return x
+    return x, y
 
 
 def randomflip_leftright(x, y=None, trans_xy=False):
@@ -234,5 +235,6 @@ def randomflip_leftright(x, y=None, trans_xy=False):
         c = tf.image.random_flip_left_right(c, seed=None)
         split_channel = int(c.shape[0] / 2)
         return c[0:split_channel, :, :], c[split_channel : c.shape[0], :, :]
-    else:
+    if y is None:
         return tf.image.random_flip_left_right(x, seed=None)
+    return tf.image.random_flip_left_right(x, seed=None), y

--- a/nobrainer/tests/test_intensity_transforms.py
+++ b/nobrainer/tests/test_intensity_transforms.py
@@ -89,7 +89,8 @@ def test_intensityMasking():
 
     y = np.random.rand(*x.shape)
     results, y_out = intensity_transforms.intensityMasking(
-        x, mask_x=mask_x, y=y, trans_xy=False)
+        x, mask_x=mask_x, y=y, trans_xy=False
+    )
     results = tf.squeeze(results)
     np.testing.assert_allclose(results.numpy(), expected)
     np.testing.assert_array_equal(y_out, y)
@@ -107,6 +108,8 @@ def test_contrastAdjust():
     np.testing.assert_allclose(expected, results.numpy(), rtol=1e-05)
 
     y = np.random.rand(*x.shape)
-    results, y_out = intensity_transforms.contrastAdjust(x, y, trans_xy=False, gamma=1.5)
+    results, y_out = intensity_transforms.contrastAdjust(
+        x, y, trans_xy=False, gamma=1.5
+    )
     np.testing.assert_allclose(expected, results.numpy(), rtol=1e-05)
     np.testing.assert_array_equal(y_out, y)

--- a/nobrainer/tests/test_intensity_transforms.py
+++ b/nobrainer/tests/test_intensity_transforms.py
@@ -1,5 +1,4 @@
 import numpy as np
-import pytest
 import tensorflow as tf
 
 from nobrainer import intensity_transforms

--- a/nobrainer/tests/test_intensity_transforms.py
+++ b/nobrainer/tests/test_intensity_transforms.py
@@ -5,7 +5,6 @@ import tensorflow as tf
 from nobrainer import intensity_transforms
 
 
-@pytest.fixture(scope="session")
 def test_addGaussianNoise():
     shape = (10, 10, 10)
     x = np.ones(shape).astype(np.float32)
@@ -27,31 +26,55 @@ def test_addGaussianNoise():
     assert y_out.shape == y.shape
     assert np.sum(noise_x - noise_y) < 1e-5
 
+    # test sending y, but not transforming it
+    x_out, y_out = intensity_transforms.addGaussianNoise(
+        x, y, trans_xy=False, noise_mean=0.0, noise_std=1
+    )
+    x_out = x_out.numpy()
+    assert x_out.shape == x.shape
+    assert np.sum(x_out - x) != 0
+    assert y_out.shape == y.shape
+    np.testing.assert_array_equal(y_out, y)
+
 
 def test_minmaxIntensityScaling():
     x = np.random.rand(10, 10, 10).astype(np.float32)
     y = np.random.randint(0, 2, size=(10, 10, 10)).astype(np.float32)
-    x, y = intensity_transforms.minmaxIntensityScaling(x, y, trans_xy=True)
-    x_out = x.numpy()
-    y_out = y.numpy()
+    x_out, y_out = intensity_transforms.minmaxIntensityScaling(x, y, trans_xy=True)
+    x_out = x_out.numpy()
+    y_out = y_out.numpy()
     assert x_out.min() - 0.0 < 1e-5
     assert y_out.min() - 0.0 < 1e-5
     assert 1 - x_out.max() < 1e-5
     assert 1 - y_out.max() < 1e-5
 
+    x_out, y_out = intensity_transforms.minmaxIntensityScaling(x, y, trans_xy=False)
+    x_out = x_out.numpy()
+    assert x_out.min() - 0.0 < 1e-5
+    assert 1 - x_out.max() < 1e-5
+    np.testing.assert_array_equal(y_out, y)
+
 
 def test_customIntensityScaling():
     x = np.random.rand(10, 10, 10).astype(np.float32)
     y = np.random.randint(0, 2, size=(10, 10, 10)).astype(np.float32)
-    x, y = intensity_transforms.customIntensityScaling(
+    x_out, y_out = intensity_transforms.customIntensityScaling(
         x, y, trans_xy=True, scale_x=[0, 100], scale_y=[0, 3]
     )
-    x_out = x.numpy()
-    y_out = y.numpy()
+    x_out = x_out.numpy()
+    y_out = y_out.numpy()
     assert x_out.min() - 0.0 < 1e-5
     assert y_out.min() - 0.0 < 1e-5
     assert 100 - x_out.max() < 1e-5
     assert 3 - y_out.max() < 1e-5
+
+    x_out, y_out = intensity_transforms.customIntensityScaling(
+        x, y, trans_xy=False, scale_x=[0, 100], scale_y=[0, 3]
+    )
+    x_out = x_out.numpy()
+    assert x_out.min() - 0.0 < 1e-5
+    assert 100 - x_out.max() < 1e-5
+    np.testing.assert_array_equal(y_out, y)
 
 
 def test_intensityMasking():
@@ -64,6 +87,13 @@ def test_intensityMasking():
     results = tf.squeeze(results)
     np.testing.assert_allclose(results.numpy(), expected)
 
+    y = np.random.rand(*x.shape)
+    results, y_out = intensity_transforms.intensityMasking(
+        x, mask_x=mask_x, y=y, trans_xy=False)
+    results = tf.squeeze(results)
+    np.testing.assert_allclose(results.numpy(), expected)
+    np.testing.assert_array_equal(y_out, y)
+
 
 def test_contrastAdjust():
     gamma = 1.5
@@ -75,3 +105,8 @@ def test_contrastAdjust():
     )
     results = intensity_transforms.contrastAdjust(x, gamma=1.5)
     np.testing.assert_allclose(expected, results.numpy(), rtol=1e-05)
+
+    y = np.random.rand(*x.shape)
+    results, y_out = intensity_transforms.contrastAdjust(x, y, trans_xy=False, gamma=1.5)
+    np.testing.assert_allclose(expected, results.numpy(), rtol=1e-05)
+    np.testing.assert_array_equal(y_out, y)

--- a/nobrainer/tests/test_spatial_transforms.py
+++ b/nobrainer/tests/test_spatial_transforms.py
@@ -17,6 +17,17 @@ def test_centercrop():
     assert x.shape[1] == fine & x.shape[0] == fine & x.shape[2] == shape[2]
     assert y.shape[1] == fine & y.shape[0] == fine & y.shape[2] == shape[2]
 
+    # Test for passing y but not transoforming it
+    shape = (10, 10, 10)
+    x = np.ones(shape).astype(np.float32)
+    y = np.random.randint(0, 2, size=shape).astype(np.float32)
+    fine = int(x.shape[1])
+    x, y_out = transformations.centercrop(x, y, trans_xy=False, finesize=fine)
+    x = x.numpy()
+    # Test for output shapes
+    assert x.shape[1] == fine & x.shape[0] == fine & x.shape[2] == shape[2]
+    np.testing.assert_array_equal(y_out, y)
+
     # test for both x and y
     shape = (10, 10, 10)
     x = np.ones(shape).astype(np.float32)
@@ -119,6 +130,12 @@ def test_spatialConstantPadding():
     np.testing.assert_allclose(x_expected, resultx.numpy())
     np.testing.assert_allclose(y_expected, resulty.numpy())
 
+    resultx, resulty = transformations.spatialConstantPadding(
+        x, y, trans_xy=False, padding_zyx=[0, 2, 2]
+    )
+    np.testing.assert_allclose(x_expected, resultx.numpy())
+    np.testing.assert_array_equal(resulty, y)
+
 
 def test_randomCrop():
     x = np.random.rand(10, 10, 10).astype(np.float32)
@@ -129,6 +146,11 @@ def test_randomCrop():
     assert np.shape(res_y.numpy()) == expected_shape
     assert np.all(np.in1d(np.ravel(res_x), np.ravel(x)))
     assert np.all(np.in1d(np.ravel(res_y), np.ravel(y)))
+
+    res_x, res_y = transformations.randomCrop(x, y, trans_xy=False, cropsize=3)
+    assert np.shape(res_x.numpy()) == expected_shape
+    assert np.all(np.in1d(np.ravel(res_x), np.ravel(x)))
+    np.testing.assert_array_equal(res_y, y)
 
 
 def test_resize():
@@ -141,6 +163,12 @@ def test_resize():
     assert np.shape(results_x.numpy()) == expected_shape
     assert np.shape(results_y.numpy()) == expected_shape
 
+    results_x, results_y = transformations.resize(
+        x, y, trans_xy=False, size=[5, 5], mode="bicubic"
+    )
+    assert np.shape(results_x.numpy()) == expected_shape
+    np.testing.assert_array_equal(results_y, y)
+
 
 def test_randomflip_leftright():
     x = np.random.rand(3, 3, 3).astype(np.float32)
@@ -149,3 +177,8 @@ def test_randomflip_leftright():
     expected_shape = (3, 3, 3)
     assert np.shape(res_x.numpy()) == expected_shape
     assert np.shape(res_y.numpy()) == expected_shape
+
+    res_x, res_y = transformations.randomflip_leftright(x, y, trans_xy=False)
+    expected_shape = (3, 3, 3)
+    assert np.shape(res_x.numpy()) == expected_shape
+    np.testing.assert_array_equal(res_y, y)

